src/device/sound.rs

```diff
@@ -0,0 +1,1563 @@
+        })?;
+        // rsp is just a header, so it can be compared with VirtIOSndHdr
+        if rsp == VirtIOSndHdr::from(RequestStatusCode::VirtioSndSOk) {
+            Ok(())
+        } else {
+            Err(Error::IoError)
+        }
+    }
+
+    /// Stop a stream with specified stream ID.
+    pub fn pcm_stop(&mut self, stream_id: u32) -> Result {
+        if !self.set_up {
+            self.set_up();
+            self.set_up = true;
+        }
+        let request_hdr = VirtIOSndHdr::from(CommandCode::VirtioSndRPcmStop);
+        let rsp: VirtIOSndHdr = self.request(VirtIOSndPcmHdr {
+            hdr: request_hdr,
+            stream_id,
+        })?;
+        // rsp is just a header, so it can be compared with VirtIOSndHdr
+        if rsp == VirtIOSndHdr::from(RequestStatusCode::VirtioSndSOk) {
+            Ok(())
+        } else {
+            Err(Error::IoError)
+        }
+    }
+
+    /// Transfer PCM frame to device, based on the stream type(OUTPUT/INPUT).
+    /// 
+    /// This is a blocking method that will not return until the audio playback is complete.
+    pub fn pcm_xfer(&mut self, stream_id: u32, frames: &[u8]) -> Result {
+        const U32_SIZE: usize = mem::size_of::<u32>();
+        if !self.set_up {
+            self.set_up();
+            self.set_up = true;
+        }
+        if !self.pcm_parameters[stream_id as usize].setup {
+            warn!("Please set parameters for a stream before using it!");
+            return Err(Error::IoError);
+        }
+        let buffer_size = self.pcm_parameters[stream_id as usize].buffer_bytes as usize;
+        assert!(buffer_size * 2 <= frames.len());
+        let mut buf1 = vec![0; U32_SIZE + buffer_size];
+        let mut buf2 = vec![0; U32_SIZE + buffer_size];
+
+        buf1[..U32_SIZE].copy_from_slice(&stream_id.to_le_bytes());
+        buf1[U32_SIZE..U32_SIZE + buffer_size].copy_from_slice(&frames[..buffer_size]);
+
+        buf2[..U32_SIZE].copy_from_slice(&stream_id.to_le_bytes());
+        buf2[U32_SIZE..U32_SIZE + buffer_size]
+            .copy_from_slice(&frames[buffer_size..buffer_size * 2]);
+
+        let mut outputs = vec![0; 32];
+        let mut token1 = unsafe { self.tx_queue.add(&[&buf1], &mut [&mut outputs]).unwrap() };
+        let mut token2 = unsafe { self.tx_queue.add(&[&buf2], &mut [&mut outputs]).unwrap() };
+
+        let xfer_times = if frames.len() % buffer_size == 0 {
+            frames.len() / buffer_size
+        } else {
+            frames.len() / buffer_size + 1
+        };
+
+        let mut turn1 = true;
+
+        for i in 2..xfer_times {
+            if self.tx_queue.should_notify() {
+                self.transport.notify(TX_QUEUE_IDX);
+            }
+
+            let start_byte = i * buffer_size;
+            let end_byte = if i != xfer_times - 1 {
+                (i + 1) * buffer_size
+            } else {
+                frames.len()
+            };
+            if turn1 {
+                while let Err(_) = unsafe {
```

Comment:
Some errors should be passed back to the caller rather than being retried forever. I suggest this loop until `self.tx_queue.can_pop()` returns true instead, and then calls `pop_used` with any error being propagated up.

Reply:
Fixed.